KAFKA-16577: New consumer fails with stop within allotted timeout in consumer_test.py system test

kirktrue opened the pull request:
The `KafkaAsyncConsumer` would occasionally fail to stop when `wakeup()` was invoked. It turns out that there's a race condition between the thread that invokes `wakeup()` and the thread that is performing an action on the consumer. If the operation's `Future` is *already* completed by thread A when thread B invoke's `completeExceptionally()` inside `wakeup()`, the `WakeupException` will be ignored. We should use the return value from `completeExceptionally()` to determine if that call actually triggered completion of the `Future`. If that method returns `false`, that signals that the `Future` was already completed, and the exception we passed to `completeExceptionally()` was ignored. Therefore, we then need to return a new `WakeupFuture` instead of `null` so that the *next* call to `setActiveTask()` will throw the `WakeupException`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kirktrue committed:
KAFKA-16577: New consumer fails with stop within allotted timeout in consumer_test.py system test

WIP

kirktrue committed:
Create test_suite.yml

kirktrue committed:
Debugging added

kirktrue committed:
Merge branch 'trunk' into KAFKA-16577-fix-ignored-wakeup

kirktrue committed:
Update VerifiableConsumer.java

kirktrue committed:
Reverting debug

kirktrue committed:
If the Future was already triggered, return WakeupException to ensure the *next* call throws the error

kirktrue committed:
Reverting unrelated changes

kirktrue committed:
Reverted debug logging

kirktrue committed:
Update WakeupTrigger.java

kirktrue committed:
Update verifiable_consumer.py

kirktrue committed:
Reverting changes

kirktrue commented:
@lucasbru/@cadonna —please review. Thanks!

kirktrue committed:
Merge branch 'trunk' into KAFKA-16577-fix-ignored-wakeup

kirktrue commented:
@cadonna—can you review this PR? Thanks!

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/WakeupTrigger.java:
```diff
@@ -43,8 +43,12 @@ public void wakeup() {
                 ActiveFuture active = (ActiveFuture) task;
-                active.future().completeExceptionally(new WakeupException());
-                return null;
+                boolean wasTriggered = active.future().completeExceptionally(new WakeupException());
```
Could we add a test in `WakeupTriggerTest` for this?

kirktrue replied:
Will do.

kirktrue replied:
I added some unit tests.

lucasbru reviewed (commented):
Thanks for the PR! Mostly looking good to me, but would be good to add a little unit test to stay consistent.

kirktrue committed:
Merge branch 'trunk' into KAFKA-16577-fix-ignored-wakeup

kirktrue committed:
Added tests

The tests ensure that when the task is completed (by normally or otherwise) between setActiveTask() and wakeup(), when maybeTriggerWakeup() is later called, it will trigger the WakeupException.

kirktrue commented:
@lucasbru—please re-review. Thanks!

lucasbru reviewed (approved):
LGTM, thanks!
